### PR TITLE
[SPARK-15678] Add support to REFRESH data source paths

### PR DESCRIPTION
**Pull request by sameeragarwal**
## What changes were proposed in this pull request?

Spark currently incorrectly continues to use cached data even if the underlying data is overwritten.

Current behavior:

``` scala
val dir = "/tmp/test"
sqlContext.range(1000).write.mode("overwrite").parquet(dir)
val df = sqlContext.read.parquet(dir).cache()
df.count() // outputs 1000
sqlContext.range(10).write.mode("overwrite").parquet(dir)
sqlContext.read.parquet(dir).count() // outputs 1000 <---- We are still using the cached dataset
```

This patch fixes this bug by adding support for `REFRESH path` that invalidates and refreshes all the cached data (and the associated metadata) for any dataframe that contains the given data source path.

Expected behavior:

``` scala
val dir = "/tmp/test"
sqlContext.range(1000).write.mode("overwrite").parquet(dir)
val df = sqlContext.read.parquet(dir).cache()
df.count() // outputs 1000
sqlContext.range(10).write.mode("overwrite").parquet(dir)
spark.catalog.refreshResource(dir)
sqlContext.read.parquet(dir).count() // outputs 10 <---- We are not using the cached dataset
```
## How was this patch tested?

Unit tests for overwrites and appends in `ParquetQuerySuite` and `CachedTableSuite`.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala [hunk 1]
```diff
@@ -157,4 +161,38 @@ private[sql] class CacheManager extends Logging {
...
+   */
+  private[sql] def invalidateCachedPath(
+      sparkSession: SparkSession, resourcePath: String): Unit = writeLock {
+    val (fs, qualifiedPath) = {
+      val path = new Path(resourcePath)
+      val fs = path.getFileSystem(sparkSession.sessionState.newHadoopConf())
+      (fs, path.makeQualified(fs.getUri, fs.getWorkingDirectory))
+    }
+    cachedData.foreach {
+      case data if data.plan.find {
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala]: Could you move this into a separate function; it was kinda hard to understand that it is a part of the `case` guard.

**Comment by hvanhovell** [comment]
Looks pretty good. Left one comment.

**Comment by sameeragarwal** [comment]
Thanks, I pulled it out in a separate function.

File: sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala [hunk 2]
```diff
@@ -226,4 +226,11 @@ abstract class Catalog {
    */
   def refreshTable(tableName: String): Unit
 
+  /**
+   * Invalidate and refresh all the cached data (and the associated metadata) for any dataframe that
+   * contains the given data source path.
+   *
+   * @since 2.0.0
+   */
+  def refreshResource(path: String): Unit
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala]: Should we call it `invalidateCache()` to reflect the things we actually done?

Also, it's a bit confusing to have this API on `catalog`, can we put it on SparkSession?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala [hunk 3]
```diff
@@ -157,4 +161,49 @@ private[sql] class CacheManager extends Logging {
...
+
+  /**
+   * Traverses a given `plan` and searches for the occurrences of `qualifiedPath` in the
+   * [[org.apache.spark.sql.execution.datasources.FileCatalog]] of any [[HadoopFsRelation]] node
+   * in the plan. If found, we refresh the metadata and return true. Otherwise, this method returns
+   * false.
+   */
+  private def lookupAndRefresh(plan: LogicalPlan, fs: FileSystem, qualifiedPath: Path): Boolean = {
+    plan match {
+      case lr: LogicalRelation => lr.relation match {
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala]: The LogicalRelation could be part of cached logical plan, should we also invalidate that?

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala]: nvm, we are using `plan.find` in invalidateCachedPath

File: sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala [hunk 4]
```diff
@@ -157,4 +161,49 @@ private[sql] class CacheManager extends Logging {
...
+      val fs = path.getFileSystem(sparkSession.sessionState.newHadoopConf())
+      (fs, path.makeQualified(fs.getUri, fs.getWorkingDirectory))
+    }
+
+    cachedData.foreach {
+      case data if data.plan.find(lookupAndRefresh(_, fs, qualifiedPath)).isDefined =>
+        val dataIndex = cachedData.indexWhere(cd => data.plan.sameResult(cd.plan))
+        if (dataIndex >= 0) {
+          cachedData(dataIndex).cachedRepresentation.cachedColumnBuffers.unpersist(blocking = true)
+          cachedData.remove(dataIndex)
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala]: Once we refresh the LogicalRelation and unpersist the cachedRepresentation, do we still need to remove it and add it back?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala [hunk 5]
```diff
@@ -157,4 +161,49 @@ private[sql] class CacheManager extends Logging {
...
+      val path = new Path(resourcePath)
+      val fs = path.getFileSystem(sparkSession.sessionState.newHadoopConf())
+      (fs, path.makeQualified(fs.getUri, fs.getWorkingDirectory))
+    }
+
+    cachedData.foreach {
+      case data if data.plan.find(lookupAndRefresh(_, fs, qualifiedPath)).isDefined =>
+        val dataIndex = cachedData.indexWhere(cd => data.plan.sameResult(cd.plan))
+        if (dataIndex >= 0) {
+          cachedData(dataIndex).cachedRepresentation.cachedColumnBuffers.unpersist(blocking = true)
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala]: data = cachedData(dataIndex), right?

File: sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala (hunk 2 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala]: I'm confusing with these Catalog/SessionCatalog/ExternalCatalog here, thought this is SessionCatalog or ExternalCatalog, so it make sense to be here (together with other API related to cache).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala (hunk 4 above)
Comment (sameeragarwal) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala]: Unfortunately, I didn't find a way around it. The old plan has an incorrect catalog, and removing and adding the new one seems like the only option.

File: sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala (hunk 2 above)
Comment (sameeragarwal) [inline sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala]: I like `invalidateCache()` but the reason for choosing `refreshResource()` was to make it sound similar to `refreshTable ()` above. Let me know if you prefer one over the other.

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala]: resource does sound a bit werid to me

Comment (sameeragarwal) [inline sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala]: alright, changed this to `refreshByPath` based on @ericl's suggestion :)

**Comment by davies** [comment]
LGTM, 
Merging this into master and 2.0, thanks!

**Pull request closed without merging**